lib/spack/spack/mirror.py

```diff
@@ -188,45 +203,90 @@ def create(path, specs, **kwargs):
+    mirror_stats = MirrorStats()
     try:
         spack.caches.mirror_cache = mirror_cache
         # Iterate through packages and download all safe tarballs for each
-        for spec in version_specs:
-            add_single_spec(spec, mirror_root, categories, **kwargs)
+        for spec in specs:
+            mirror_stats.next_spec(spec)
+            add_single_spec(spec, mirror_root, mirror_stats)
     finally:
         spack.caches.mirror_cache = None
 
-    categories['mirrored'] = list(mirror_cache.new_resources)
-    categories['present'] = list(mirror_cache.existing_resources)
+    return mirror_stats.stats()
+
 
-    return categories['present'], categories['mirrored'], categories['error']
+class MirrorStats(object):
+    def __init__(self):
+        self.present = {}
+        self.new = {}
+        self.errors = set()
 
+        self.current_spec = None
+        self.added_resources = set()
+        self.existing_resources = set()
 
-def add_single_spec(spec, mirror_root, categories, **kwargs):
+    def next_spec(self, spec):
+        self._tally_current_spec()
+        self.current_spec = spec
+
+    def _tally_current_spec(self):
+        if self.current_spec:
+            if self.added_resources:
+                self.new[self.current_spec] = len(self.added_resources)
+            if self.existing_resources:
+                self.present[self.current_spec] = len(self.existing_resources)
+            self.added_resources = set()
+            self.existing_resources = set()
+        self.current_spec = None
+
+    def stats(self):
+        self._tally_current_spec()
+        return list(self.present), list(self.new), list(self.errors)
+
+    def already_existed(self, resource):
+        # If an error occurred after caching a subset of a spec's
+        # resources, a secondary attempt may consider them already added
+        if resource not in self.added_resources:
+            self.existing_resources.add(resource)
+
+    def added(self, resource):
+        self.added_resources.add(resource)
+
+    def error(self):
+        self.errors.add(self.current_spec)
+
+
+def add_single_spec(spec, mirror_root, mirror_stats):
     tty.msg("Adding package {pkg} to mirror".format(
         pkg=spec.format("{name}{@version}")
     ))
-    try:
-        spec.package.do_fetch()
-        spec.package.do_clean()
-
-    except Exception as e:
-        tty.debug(e)
+    num_retries = 3
+    while num_retries > 0:
+        try:
+            with spec.package.stage as pkg_stage:
+                pkg_stage.cache_mirror(mirror_stats)
+                for patch in spec.package.all_patches():
+                    patch.fetch(pkg_stage)
+                    if patch.cache():
+                        patch.cache().cache_mirror(mirror_stats)
+                    patch.clean()
+            exception = None
+            break
+        except Exception as e:
+            exc_tuple = sys.exc_info()
+            exception = e
+        num_retries -= 1
```

Comment:
An optional modification you could make here would be to notice if you get a 404 (not found) and give up on retrying then, as you likely will not find the thing on retry.  There are probably other errors that we should just give up on.  This can be done in the future as it's pretty minor, but it would be nice to propagate thie response back up to the top-level summary error messages.

Reply:
I think trying 3 404's is not much more taxing than 1. However, knowing that there was a 404 is probably useful because that's the case where the user should actually update the URL in the `package.py` file. However I think that improving that could be done later.